background/index.js

```diff
@@ -17,13 +17,171 @@
 /* global chrome */
 
 import Processor from './processor';
+import Ws from './ws';
 
-const processor = new Processor();
+import { UI, TRANSPORT_UNINITIALIZED } from '../shared';
 
 chrome.runtime.onConnect.addListener((port) => {
-  console.assert(port.name === 'id');
+  if (port.name === 'secureApi') {
+    port.onMessage.addListener(web3Message(port));
+    return;
+  }
+
+  if (port.name === 'barScripts') {
+    port.onMessage.addListener(loadScripts(port));
+    return;
+  }
+
+  // TODO [ToDr] Use connection without elevated privileges here!
+  if (port.name === 'web3') {
+    port.onMessage.addListener(web3Message(port));
+    return;
+  }
+
+  if (port.name === 'id') {
+    port.onMessage.addListener(processId(port));
+    return;
+  }
+
+  throw new Error(`Unrecognized port: ${port.name}`);
+});
+
+let openedTabId = null;
+let transport = null;
+// Attempt to extract token on start if not available.
+extractToken();
+
+chrome.runtime.onMessage.addListener((request, sender, callback) => {
+  if (!(transport && transport.isConnected) && request.token) {
+    if (transport) {
+      // TODO [ToDr] kill old transport!
+    }
+
+    if (openedTabId) {
+      chrome.tabs.remove(openedTabId);
```

Comment:
Agreed with the welcome message, as per other extensions. Would log it in another issue though.
Because even if the tab is open in background, you can still see it open and close, right?